tap_mailchimp/streams.py

```diff
@@ -0,0 +1,538 @@
+import json
+import time
+import random
+import tarfile
+
+import singer
+from singer import metrics, metadata, Transformer
+from singer.utils import strptime_to_utc, should_sync_field
+from requests.exceptions import HTTPError
+
+LOGGER = singer.get_logger()
+
+MIN_RETRY_INTERVAL = 2  # 2 seconds
+MAX_RETRY_INTERVAL = 300  # 5 minutes
+MAX_RETRY_ELAPSED_TIME = 43200  # 12 hours
+
+# Break up reports_email_activity batches to iterate over chunks
+EMAIL_ACTIVITY_BATCH_SIZE = 100
+
+DEFAULT_PAGE_SIZE = 1000
+
+
+class BatchExpiredError(Exception):
+    pass
+
+
+def next_sleep_interval(previous_sleep_interval):
+    """Function to send the time to sleep based on previous sleep interval"""
+    min_interval = previous_sleep_interval or MIN_RETRY_INTERVAL
+    max_interval = previous_sleep_interval * 2 or MIN_RETRY_INTERVAL
+    return min(MAX_RETRY_INTERVAL, random.randint(min_interval, max_interval))
+
+
+def chunk_campaigns(sorted_campaigns, chunk_bookmark):
+    """Function to break list for sorted campaigns into the batch size and send the chunked list"""
+    chunk_start = chunk_bookmark * EMAIL_ACTIVITY_BATCH_SIZE
+    chunk_end = chunk_start + EMAIL_ACTIVITY_BATCH_SIZE
+
+    if chunk_bookmark > 0:
```

Comment:
Def of chunk_bookmark is unclear, please add a code comment here to explain the same and this condition.

Reply:
Added code comments.